Support attaching to an existing request queue in HTTP.SYS

bkatms opened the pull request:
Support attaching to an existing request queue in HTTP.SYS ( #5866).  This is an addendum to 
“https://github.com/aspnet/AspNetCore/issues/13461”.  

This is criticial (P0) functionality required by Office 365 in support of our new ASP.Net Core 3.X architecture for data access to Substrate Storage.  This is the storage that is mission critical for Office 365.

Tratcher committed:
Add sample app for manual testing

bkatms committed:
Initial port of old branch for existing queues #5866

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/aspnet/AspNetCore?pullRequest=14182) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/aspnet/AspNetCore?pullRequest=14182) before we can accept your contribution.<br/><br/>:x: bkatms [sign now](https://cla.dotnetfoundation.org/aspnet/AspNetCore?pullRequest=14182)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/aspnet/AspNetCore?pullRequest=14182) it.</sub>

Tratcher reviewed (commented):
Thanks, getting this rebased does help.

There are some open issues from #5866 that we hadn't figured out how to address yet. Most of these are centered around ensuring each listener on a queue is using the same configuration. For things like timeouts it's at most a warning that the config is out of sync, but for auth and PathBase it can affect how requests are processed. 
- PathBase will need a fallback mode when attached to existing queues where it can't look things up by index.
- For auth we might want to query the queue settings to see if they match and fail otherwise
- I'm not sure if we can query for the timeout settings.
- Some options can be changed after the server starts such as UrlPrefixes. We need to make sure those code paths are disabled (and throw?) if this instance did not create the queue.
- Do we need to surface an API that says if this instance did attach or create?

I'll pull your branch and see what I can work on in your off hours. Want to give me push access to your fork?

anurse commented:
FYI @Tratcher check in with me before merging. As this is feature work we'll require special approval for 3.1 (shouldn't be a problem).

Tratcher commented:
Didn't make any progress on this today other than squashing and relocating my old branch to make it easier to reference: https://github.com/aspnet/AspNetCore/commit/2c4824398786964af7bb0a0a470235946cab6de9

bkatms commented on src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs:
```diff
@@ -14,26 +14,66 @@ internal class RequestQueue
+                    HttpApi.Version,
+                    requestQueueName,
+                    null,
+                    (uint) flags,
```
In our case we will have an overall management process doing the main parent Q configuration and setting the settings on behalf of the ASP.Net Core processes as part of the integration into our routing fabric so I was worried less about that.  I can explain this more offline to you if you are interested in the full scenario in our our case.

We will not be changing UrlPrefixes at this time so we don't need that support and throwing is fine.

We don't care if we created ourselves or not,  HTTP.SYS will get rid of the Q when the last reference to it is gone.  However, I can see if you might want to return a flag.

Typically I assume that the process will not be able to attach to the named Q in the first place if it does not have permissions to do so. 

Thanks!

Tratcher committed:
Pull in other parts of old PR (Some tests failing)

Tratcher commented:
Made some progress, I pulled in the missing parts from the prior branch including tests and some misc fixes. I'll start on the actual review tomorrow.

davidfowl commented on src/Servers/HttpSys/ref/Microsoft.AspNetCore.Server.HttpSys.netcoreapp.cs:
```diff
@@ -51,12 +51,20 @@ public partial class HttpSysOptions
+    public enum RequestQueueMode
+    {
+        Create = 0,
+        Controller = 1,
```
For my edification, what’s controller?

Tratcher replied:
It's an orchestration role, the controller creates and configures the queues, but does not listen for requests. The customer already has one of these written in C++. We may consider exposing something like that eventually, but it would be separate from the web app model and wouldn't be an IServer.

Tratcher committed:
Fix path base, get tests passing

Tratcher committed:
Ref assemblies

# Conflicts:
#	src/Servers/HttpSys/ref/Microsoft.AspNetCore.Server.HttpSys.netcoreapp.cs

Tratcher committed:
Remove controller mode

Tratcher committed:
Explicit flags type

Tratcher committed:
Rename Mode property. Add comments about what's settings apply.

Tratcher commented on src/Servers/HttpSys/src/RequestProcessing/Request.cs:
```diff
@@ -66,19 +67,26 @@ internal Request(RequestContext requestContext, NativeRequestContext nativeReque
             }
-            else
+            // Make sure it was really a match. The queue configuration may be out of sync with other instances.
+            else if (prefix.PathWithoutTrailingSlash.Length > 0
```
Nevermind, this logic is wrong. We can only rely on UrlContext if this instance created the queue. Otherwise we don't know what those indexes were set to. We'll have to iterate through the collection and look for the longest match manually.

Tratcher committed:
PathBase searching

Tratcher committed:
Reorder CreateOrAttach

Tratcher committed:
Remove outdated test

Tratcher committed:
Do not add a default address

Tratcher committed:
PR Cleanup

Tratcher commented:
I think that's everything for now.
- I've dealt with PathBase. It can't query prefixes configured elsewhere but can do a match against ones you provide in the app.
- For auth you must make sure your schemes are in sync or it's not going to work well. We can discuss guardrails in the future if people have issues.
- Many settings are marked as not applicable in Attach mode.
- If using CreateOrAttach, there is a log statement indicating if it attached but there's no API to show the results in the app. This isn't a priority for 3.1 since our driving customer doesn't plan to use CreateOrAttach. In the future consider surfacing this as a server feature? I'm not sure what scenarios would require knowing the difference (HotAdd maybe?).
- Suppressed default urls when attached to an existing queue.

@halter73 @bkatms please review.

anurse commented:
paging @davidfowl @rynowak for API review (not sure who else wants to be involved or if a GitHub team has been set up yet). 

The API here is super minimal, just [a new option that has an enum value](https://github.com/aspnet/AspNetCore/pull/14182/files#diff-3c0a65565dfe2587d01d4adebc2d2d02). LGTM.

davidfowl commented on src/Servers/HttpSys/src/RequestQueueMode.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace Microsoft.AspNetCore.Server.HttpSys
+{
+    public enum RequestQueueMode
```
Add a doc comment for the enum

davidfowl commented on src/Servers/HttpSys/src/UrlPrefixCollection.cs:
```diff
@@ -57,10 +59,34 @@ internal UrlPrefix GetPrefix(int id)
             }
         }
 
+        internal bool TryMatchLongestPrefix(bool isHttps, string host, string originalPath, out string pathBase, out string remainingPath)
```
Does this get called per request?

Tratcher replied:
Yes, but the list should be short.

bkatms commented:
The changes look fine to me.  Thank you.

bkatms commented:
Note relative to "Created" - my understanding is that HTTP.SYS will automatically delete a queue if there are no active references to it by any process.

In our case we will have a management process create and configure the queues and pass then name into our ASP.NET Core process - this is similar to how IIS works when it starts W3WP.exe.  We will either directly spawn (security context inheritance) or set the ACLs appropriately - TBD.  In any case we plan on handling that part.

Let me know when you have a drop for me to begin testing on.

Cheers!

Tratcher committed:
Doc comments

jkotalik commented on src/Servers/HttpSys/src/AuthenticationManager.cs:
```diff
@@ -32,6 +32,9 @@ internal AuthenticationManager()
         }
 
+        /// <summary>
+        /// When attaching to an existing queue this setting must match the one used to create the queue.
```
This should probably be a remark rather than summary.

Tratcher replied:
Remarks aren't visible in most situations.

jkotalik commented on src/Servers/HttpSys/src/HttpSysOptions.cs:
```diff
@@ -38,7 +38,7 @@ public string RequestQueueName
             set
             {
-                if (value.Length > MaximumRequestQueueNameLength)
+                if (value?.Length > MaximumRequestQueueNameLength)
```
Was this just a random find?

Tratcher replied:
RequestQueueName was added as part of this feature, just in an earlier PR. And yeah, when writing tests I accidentally tried setting it to null and it threw null ref.

jkotalik commented on src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs:
```diff
@@ -14,20 +15,54 @@ internal class RequestQueue
+                        out requestQueueHandle);
+            }
+
+            if (flags == HttpApiTypes.HTTP_CREATE_REQUEST_QUEUE_FLAG.OpenExisting && statusCode == UnsafeNclNativeMethods.ErrorCodes.ERROR_FILE_NOT_FOUND)
```
Shouldn't this technically check CreateOrAttach and OpenExisting?

Tratcher replied:
That's why I'm checking the flags and not the enum.

jkotalik replied:
> For CreateOrAttach that should only happen with a strange race condition.

Yeah, the strange race condition is what I was referring to.

halter73 replied:
I don't think it's really worth guarding against that race condition, since to do so would require some sort of looping.

jkotalik commented on src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs:
```diff
@@ -39,18 +74,30 @@ internal RequestQueue(UrlGroup urlGroup, string requestQueueName, ILogger logger
                     UnsafeNclNativeMethods.FileCompletionNotificationModes.SkipCompletionPortOnSuccess |
                     UnsafeNclNativeMethods.FileCompletionNotificationModes.SkipSetEventOnHandle))
             {
+                requestQueueHandle.Dispose();
```
Why was this added?

Tratcher replied:
Eager clean up. If this secondary API call fails then the requestQueueHandle would only be cleaned up by its finalizer. The process is probably going to shut down anyways, but it's best to clean things up as locally as possible.

jkotalik commented on src/Servers/HttpSys/src/UrlPrefix.cs:
```diff
@@ -15,8 +15,10 @@ private UrlPrefix(bool isHttps, string scheme, string host, string port, int por
+            HostAndPort = string.Format(CultureInfo.InvariantCulture, "{0}:{1}", Host, Port);
             PortValue = portValue;
             Path = path;
+            PathWithoutTrailingSlash = Path.Length > 1 ? Path[0..^1] : string.Empty;
```
Get out of here with your dark magic.

Tratcher replied:
VS wrote that for me 😁

jkotalik commented on src/Servers/HttpSys/src/UrlPrefix.cs:
```diff
@@ -15,8 +15,10 @@ private UrlPrefix(bool isHttps, string scheme, string host, string port, int por
             Scheme = scheme;
             Host = host;
             Port = port;
+            HostAndPort = string.Format(CultureInfo.InvariantCulture, "{0}:{1}", Host, Port);
```
To confirm, UrlPrefixes are created once, not once per request?

Tratcher replied:
Right, these are server level config.

jkotalik commented on src/Servers/HttpSys/src/UrlPrefixCollection.cs:
```diff
@@ -57,10 +59,34 @@ internal UrlPrefix GetPrefix(int id)
+            {
+                foreach (var prefix in _prefixes.Values)
+                {
+                    if (isHttps == prefix.IsHttps
```
Add a comment with an example, especially for the longest match cases.

jkotalik commented on src/Servers/HttpSys/test/FunctionalTests/Listener/ServerOnExistingQueueTests.cs:
```diff
@@ -0,0 +1,229 @@
+        [InlineData("/base path/", "/base%20path/sub%20path", "/base path", "/sub path")]
+        [InlineData("/base葉path/", "/base%E8%91%89path/sub%E8%91%89path", "/base葉path", "/sub葉path")]
+        [InlineData("/basepath/", "/basepath/sub%2Fpath", "/basepath", "/sub%2Fpath")]
+        public async Task Server_PathSplitting(string pathBase, string requestPath, string expectedPathBase, string expectedPath)
```
Add a test for longest match?

Tratcher committed:
Longest match test

anurse commented:
@davidfowl this needs to be merged Very Soon. What if anything do you want to do for API review here?

davidfowl commented:
> For auth you must make sure your schemes are in sync or it's not going to work well. We can discuss guardrails in the future if people have issues.

What’s the failure mode here?

Tratcher commented:
> > For auth you must make sure your schemes are in sync or it's not going to work well. We can discuss guardrails in the future if people have issues.
> 
> > Many settings are marked as not applicable in Attach mode.
> 
> What’s the failure mode here?

The managed layer enforces AllowAnonymous and adds challenges based on what's configured. Worst case you end up requiring more auth than the underlying queue supports and the client gets rejected with 401s.

bkatms commented:
I am good here.

Cheers, Brian
[he/him]

From: Andrew Stanton-Nurse <notifications@github.com>
Sent: Tuesday, October 1, 2019 9:18 AM
To: aspnet/AspNetCore <AspNetCore@noreply.github.com>
Cc: Brian Kress <Brian.Kress@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [aspnet/AspNetCore] Support attaching to an existing request queue in HTTP.SYS (#14182)


@davidfowl<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdavidfowl&data=02%7C01%7CBrian.Kress%40microsoft.com%7C443d473c952345893b6908d7468af773%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637055434976760183&sdata=3woC1wVs0%2FL9iYRk%2BZwOlNMZovUBpxTR23XD3epqQ0Q%3D&reserved=0> this needs to be merged Very Soon. What if anything do you want to do for API review here?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Faspnet%2FAspNetCore%2Fpull%2F14182%3Femail_source%3Dnotifications%26email_token%3DAKV7FBDZVALZLQ47VOUOEOLQMNZ4PA5CNFSM4IYT4RY2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAB273I%23issuecomment-537112557&data=02%7C01%7CBrian.Kress%40microsoft.com%7C443d473c952345893b6908d7468af773%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637055434976770180&sdata=YdR1Kv5SpeyfQnmSgPIlXFI5UPa4QIB2PK%2Fd2oZSLzM%3D&reserved=0>, or mute the thread<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKV7FBHRS2QAEKTI7ZFCPJDQMNZ4PANCNFSM4IYT4RYQ&data=02%7C01%7CBrian.Kress%40microsoft.com%7C443d473c952345893b6908d7468af773%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637055434976770180&sdata=f5I2gcY2uV117fis4ZtXoITXLghNR5hdiF7FVeu0byw%3D&reserved=0>.

halter73 commented on src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs:
```diff
@@ -39,18 +74,30 @@ internal RequestQueue(UrlGroup urlGroup, string requestQueueName, ILogger logger
+
+            if (!Created)
+            {
+                _logger.LogInformation("Attached to an existing request queue.");
```
Should this log include the queue name?

Tratcher committed:
Include queue name in log

Tratcher commented:
@anurse this is ready to merge.
